src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,344 @@
+                         chunk_id_right);
+
+          // set range for second half to all values greater than the search value
+          range_begin = index->upper_bound(comp_values);
+          range_end = index->cend();
+          break;
+        }
+        case PredicateCondition::GreaterThan: {
+          range_begin = index->cbegin();
+          range_end = index->lower_bound(comp_values);
+          break;
+        }
+        case PredicateCondition::GreaterThanEquals: {
+          range_begin = index->cbegin();
+          range_end = index->upper_bound(comp_values);
+          break;
+        }
+        case PredicateCondition::LessThan: {
+          range_begin = index->upper_bound(comp_values);
+          range_end = index->cend();
+          break;
+        }
+        case PredicateCondition::LessThanEquals: {
+          range_begin = index->lower_bound(comp_values);
+          range_end = index->cend();
+          break;
+        }
+        default:
+          Fail("Unsupported comparison type encountered");
+      }
+
+      append_matches(range_begin, range_end, left_value.chunk_offset(), left_matches, chunk_id_left, chunk_id_right);
+    }
+  } else {
+    // No index so we fall back on a nested loop join
+    for (; left_it != left_end; ++left_it) {
+      const auto left_value = *left_it;
+      if (left_value.is_null()) continue;
+
+      for (auto right_it = right_begin; right_it != right_end; ++right_it) {
+        const auto right_value = *right_it;
+        if (right_value.is_null()) continue;
+
+        if (func(left_value.value(), right_value.value())) {
+          _pos_list_left->emplace_back(RowID{chunk_id_left, left_value.chunk_offset()});
+          _pos_list_right->emplace_back(RowID{chunk_id_right, right_value.chunk_offset()});
+
+          if (_mode == JoinMode::Left || _mode == JoinMode::Outer) {
+            left_matches[left_value.chunk_offset()] = true;
+          }
+
+          if (_mode == JoinMode::Outer || _mode == JoinMode::Right) {
+            DebugAssert(chunk_id_right < _right_in_table->chunk_count(), "invalid chunk_id in join_index");
+            DebugAssert(right_value.chunk_offset() < _right_in_table->get_chunk(chunk_id_right)->size(),
+                        "invalid chunk_offset in join_index");
+            _right_matches[chunk_id_right][right_value.chunk_offset()] = true;
+          }
+        }
+      }
+    }
+  }
+}
+
+void JoinIndex::append_matches(const BaseIndex::Iterator& range_begin, const BaseIndex::Iterator& range_end,
+                               const ChunkOffset chunk_offset_left, std::vector<bool>& left_matches,
+                               const ChunkID chunk_id_left, const ChunkID chunk_id_right) {
+  auto num_right_matches = std::distance(range_begin, range_end);
+
+  if (num_right_matches > 0) {
+    // Remember the matches for outer joins
+    if (_mode == JoinMode::Left || _mode == JoinMode::Outer) {
+      left_matches[chunk_offset_left] = true;
```

Comment:
Nitpick:
I think it's slightly confusing for `left_matches` and `_right_matches` to have different types ("depths") here. Maybe add a `auto &right_matches = _right_matches[chunk_id_right]` variable to make the uses of matches equivalent in this function?

Reply:
This confused me too.